src/Language/PureScript/TypeChecker/Roles.hs

```diff
@@ -43,23 +43,40 @@ instance Monoid RoleMap where
   mempty =
     RoleMap M.empty
 
+type RoleEnv = M.Map (Qualified (ProperName 'TypeName)) [Role]
+
+getRoleEnv :: Environment -> RoleEnv
+getRoleEnv env =
+  flip M.mapMaybe (types env) $ \case
+    (_, DataType args _) ->
+      Just $ map (\(_, _, role) -> role) args
+    (_, ExternData roles) ->
+      Just roles
+    _ ->
+      Nothing
+
+updateRoleEnv
+  :: Qualified (ProperName 'TypeName)
+  -> [Role]
+  -> RoleEnv
+  -> (Any, RoleEnv)
+updateRoleEnv qualTyName roles' roleEnv =
+  let roles = fromMaybe (repeat Phantom) $ M.lookup qualTyName roleEnv
+      mostRestrictiveRoles = zipWith min roles roles'
+      didRolesChange = any (uncurry (<)) $ zip mostRestrictiveRoles roles
+  in (Any didRolesChange, M.insert qualTyName mostRestrictiveRoles roleEnv)
+
 -- |
 -- Lookup the roles for a type in the environment. If the type does not have
 -- roles (e.g. is a type synonym or a type variable), then this function
 -- returns an empty list.
 --
-lookupEnvRoles
+lookupRoles
   :: Environment
   -> Qualified (ProperName 'TypeName)
   -> [Role]
-lookupEnvRoles env tyName =
-  case fmap snd $ M.lookup tyName (types env) of
-    Just (DataType args _) ->
-      map (\(_, _, role) -> role) args
-    Just (ExternData roles) ->
-      roles
-    _ ->
-      []
+lookupRoles env tyName =
+  fromMaybe [] $ M.lookup tyName (getRoleEnv env)
```

Comment:
This change seems a little excessive. Whenever you want to lookup a role, instead of looking it up directly, you are mapping over the entire type environment to build another environment that is a subset of the original. I understand that there _may_ be some clever laziness involved (it isn't clear to me that you are getting that with `Map` vs something like `List`), but this seems like an easy way to also accidentally introduce poor asymptotics. What was the motivation for this change?

Reply:
I wanted to share logic and thought the whole map wouldn’t actually be built thanks to laziness but I should have verified this assumption. I’ve reused a `typeKindRoles` function in both `getRoleEnv` and `lookupRoles`.